src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import hashlib
+from dataclasses import dataclass
+
+from pants.backend.python.util_rules.interpreter_constraints import InterpreterConstraints
+from pants.util.ordered_set import FrozenOrderedSet
+
+BEGIN_LOCKFILE_HEADER = b"# --- BEGIN PANTS LOCKFILE METADATA: DO NOT EDIT OR REMOVE ---"
+END_LOCKFILE_HEADER = b"# --- END PANTS LOCKFILE METADATA ---"
+
+
+@dataclass
+class LockfileMetadata:
+    invalidation_digest: str | None
+
+
+def invalidation_digest(
+    requirements: FrozenOrderedSet[str], interpreter_constraints: InterpreterConstraints
+) -> str:
+    """Returns an invalidation digest for the given requirements and interpreter constraints."""
+    m = hashlib.sha256()
+    for requirement in requirements:
+        m.update(requirement.encode("utf-8"))
+    for constraint in interpreter_constraints:
+        m.update(str(constraint).encode("utf-8"))
+    return m.hexdigest()
```

Comment:
What about platforms? Those definitely affect a resolve and so must affect any valid lockfile.

Reply:
Per #12415, `platforms` is out of scope for now. Can we discuss over on that ticket to confirm?

Reply:
Aha. Ok - that's odd. I commented over there.